cmake/templates/python_win32_wrapper.cpp

```diff
@@ -0,0 +1,177 @@
+//   disclaimer in the documentation and/or other materials provided
+//   with the distribution.
+// * Neither the name of Yujin Robot nor the names of its
+//   contributors may be used to endorse or promote products derived
+//   from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+// COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+// BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+// ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+
+/*
+ * @file python_win32_wrapper.cpp
+ *
+ * @brief Launch a Python script located in the same location and have the same name as the compiled executable
+ *
+ * @date March 2011
+ */
+
+#ifdef _WIN32
+
+#include <iostream>
+#include <windows.h>
+
+int main(int argc, char* argv[]) try
+{
+    const auto GetCurrentModuleName = []() -> std::string
+    {
+        // initialize buffer string with at least one character
+        std::string moduleName = " ";
+        while (true)
+        {
+            // retrieves the path of the executable file of the current process
+            auto result = ::GetModuleFileName(nullptr, &moduleName[0], static_cast<DWORD>(moduleName.size()));
+            if (!result)
+            {
+                throw ::GetLastError();
+            }
+            else if (result == moduleName.size() && ::GetLastError() == ERROR_INSUFFICIENT_BUFFER)
+            {
+                // buffer not large enough
+                moduleName.resize(moduleName.size() * 2);
+                continue;
+            }
+            moduleName.resize(result);
+            break;
+        }
+        return moduleName;
+    };
+
+    const auto FindPythonScript = [](const std::string& exeName) -> std::string
+    {
+        // this could become more fail-proof and readable with https://en.cppreference.com/w/cpp/filesystem from c++20
+        const std::string exeExtension = ".exe";
+        if (exeName.size() <= exeExtension.size() || exeName.substr(exeName.size() - exeExtension.size()) != exeExtension)
+        {
+            throw L"Invalid name.";
+        }
+
+        // use quoted string to ensure the correct path is used
+        return "\"" + exeName.substr(0, exeName.size() - exeExtension.size()) + "\"";
+    };
+
+    const auto GetPythonExecutable = []() -> std::string
+    {
+        std::string pythonExecutable = "python";
```

Comment:
We can't hard code the Python executable name here. It should be the one used during the CMake run since the user might have overridden it.

Reply:
> We can't hard code the Python executable name here. It should be the one used during the CMake run since the user might have overridden it.

I am not sure how well this is going to work since our build environment would have its own "overridden" Python environment, and when the user installs ros-melodic-<distribution>, such Python environment would be meaningless if they are not doing the exact same setup.

Is it okay to follow this logic:
1. try to execute script with Python interpreter set during configure time
2. if the specified Python executable is not found, fall back to (try to) use Python interpreter from system environment

Reply:
> Is it okay to follow this logic:

That sounds good to me.